### PR TITLE
Adjust scope of locals declared within a Yield Return and Local Declaration Statements according to the latest design.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #12597.

@dotnet/roslyn-compiler, @gafter Please review.

### CODE REVIEW

**Comment by jcouv** [comment]
It looks like your test failure is due to me. I'm not sure how that happened (I only checked in after all-green). Investigating.

**Comment by AlekseyTs** [comment]
@jcouv I didn't have your changes locally, it is possible new diagnostics is reported due to my changes. Let me know.

**Comment by jcouv** [comment]
@AlekseyTs My other PR https://github.com/dotnet/roslyn/pull/13258 (fixing a few small tuple bugs) had this error too, so it's not just your change. This is happening because one of my merges adds return-types to d-assignments, and the other merge added a test for Pia including d-assignment.

The unittest passes after this change: https://github.com/dotnet/roslyn/pull/13258/commits/9c41013dc411f7948f34bd46f68b96100ac63c82

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -196,8 +196,7 @@ private BoundStatement BindFixedStatementParts(FixedStatementSyntax node, Diagno
 
         private BoundStatement BindYieldReturnStatement(YieldStatementSyntax node, DiagnosticBag diagnostics)
         {
-            var binder = GetBinder(node);
-            Debug.Assert(binder != null);
+            var binder = this;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Consider inlining `this` or removing entirely.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I left it like this intentionally in order to avoid unnecessary churn in the function.

**Comment by AlekseyTs** [comment]
Closing and re-opening in order to pick up latest changes in master. This should address test failures.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs [hunk 2]
```diff
@@ -1999,7 +1999,11 @@ unsafe static void M(int[] arg)
             CreateCompilationWithMscorlib(source, options: TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
                 // (7,32): error CS1023: Embedded statement cannot be a declaration or labeled statement
                 //         fixed (int* ptr = arg) object o = null;
-                Diagnostic(ErrorCode.ERR_BadEmbeddedStmt, "object o = null;"));
+                Diagnostic(ErrorCode.ERR_BadEmbeddedStmt, "object o = null;").WithLocation(7, 32),
+                // (7,39): warning CS0219: The variable 'o' is assigned but its value is never used
+                //         fixed (int* ptr = arg) object o = null;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "o").WithArguments("o").WithLocation(7, 39)
+                );
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs]: Why the new warning here?  This seems somewhat unrelated to the change.

File: src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs [hunk 3]
```diff
@@ -181,18 +181,26 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
...
                             var decl = (LocalDeclarationStatementSyntax)innerStatement;
                             LocalDeclarationKind kind = decl.IsConst ? LocalDeclarationKind.Constant : LocalDeclarationKind.RegularVariable;
                             foreach (var vdecl in decl.Declaration.Variables)
                             {
-                                var localSymbol = MakeLocal(decl.Declaration, vdecl, kind);
+                                var localSymbol = MakeLocal(decl.Declaration, vdecl, kind, enclosingBinder);
                                 locals.Add(localSymbol);
+
+                                var value = vdecl.Initializer?.Value;
+                                if (value != null)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs]: `FindExpressionVariables` already exits early if its third parameter is `null`. Consider removing conditional?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 4]
```diff
@@ -2904,7 +2907,10 @@ public static void Bug1098197_02()
     Diagnostic(ErrorCode.ERR_InvalidExprTerm, ")").WithArguments(")").WithLocation(2, 16),
     // (2,18): error CS1023: Embedded statement cannot be a declaration or labeled statement
     // void f() { if () const int i = 0; }
-    Diagnostic(ErrorCode.ERR_BadEmbeddedStmt, "const int i = 0;").WithLocation(2, 18)
+    Diagnostic(ErrorCode.ERR_BadEmbeddedStmt, "const int i = 0;").WithLocation(2, 18),
+    // (2,28): warning CS0219: The variable 'i' is assigned but its value is never used
+    // void f() { if () const int i = 0; }
+    Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "i").WithArguments("i").WithLocation(2, 28)
                 );
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: Hmm, this looks to be a back compat breaking change at a glance: new warning for existing code. Is there a subtle twist I'm missing here that makes this back compat safe?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 5]
```diff
@@ -67,11 +67,23 @@ internal class SourceLocalSymbol : LocalSymbol
...
+        /// <summary>
+        /// Binder that should be used to bind type syntax for the local.
+        /// </summary>
+        internal Binder TypeSyntaxBinder
+        {
+            get { return _scopeBinder; } // Scope binder should be good enough for this.
+        }
+
+
         public static SourceLocalSymbol MakeForeachLocal(
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Nit: double blank lines above.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Sure, I'll remove in the next change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: The warning is reported for declarations within embedded statements. It is an error to use local declaration statement as an embedded statement, see the error above the warning. Therefore, the warning is not a breaking change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs]: I was puzzled at first too. The change ensures that local declaration statement used as an embedded statement always gets surrounded with a BoundBlock that represents a scope for the locals declared by the statement. Before, we had inconsistency, LocalBinderFactory would create a binder for the scope, but Bonder.BindPossibleEmbeddedStatement wouldn't create a block for that scope. So the bound tree was inconsistent - scope owner for the locals was not present in the tree. The change fixes this inconsistency. As a result, flow analysis detects declarations and properly tracks their scope.

File: src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs]: Will do. Thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 6]
```diff
@@ -7505,6 +7536,103 @@ static bool TakeOutParam(object y, out int x)
         }
 
         [Fact]
+        public void Scope_LocalDeclarationStmt_06()
+        {
+            var source =
+@"
+public class X
+{
+    public static void Main()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Could remove `Main` here and a few instances below.

**Comment by jcouv** [comment]
LGTM :shipit:

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs]: > I was puzzled at first too.

Most terrifying statement I've read today ;) 

That's an interesting effect of the change.

**Comment by jaredpar** [comment]
👍

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 7]
```diff
@@ -67,11 +67,23 @@ internal class SourceLocalSymbol : LocalSymbol
...
+        internal Binder ScopeBinder
         {
-            get { return binder; }
+            get { return _scopeBinder; }
         }
 
+        /// <summary>
+        /// Binder that should be used to bind type syntax for the local.
+        /// </summary>
+        internal Binder TypeSyntaxBinder
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: >TypeSyntaxBinder [](start = 24, length = 16)

Yesterday I discussed with you a number of refactorings I was doing around here, and you seem to now be doing some of them yourself, thereby inducing merge conflicts. Is there a reason this was needed in the context of these scope changes?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 8]
```diff
@@ -463,43 +492,7 @@ private sealed class LocalWithInitializer : SourceLocalSymbol
...
-                    parentBinder = parentBinder.Next;
-                }
-
-                Debug.Assert(parentBinder != null);
-#endif 
-
-                var newBinder = initializerBinder ?? this.binder;
-                var initializerOpt = newBinder.BindInferredVariableInitializer(diagnostics, RefKind, _initializer, _initializer);
+                var initializerOpt = (this._initializerBinderOpt ?? this.ScopeBinder).BindInferredVariableInitializer(diagnostics, RefKind, _initializer, _initializer);
                 return initializerOpt?.Type;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Thank you!

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: It wasn't absolutely necessary, but felt appropriate to me, given the nature of the other changes I was making. Plus I didn't think you had any changes in flight. Sorry about the inconvenience this might have caused you. I hope the conflicts wouldn't be hard to deal with.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: I think the end result will be a huge improvement.

---
In reply to: [75558031](https://github.com/dotnet/roslyn/pull/13271#discussion_r75558031) [](ancestors = 75558031)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 9]
```diff
@@ -2649,6 +2649,9 @@ static bool TakeOutParam(object y, out bool x)
                 // (98,29): error CS0103: The name 'y12' does not exist in the current context
                 //         while (TakeOutParam(y12, out var x12));
                 Diagnostic(ErrorCode.ERR_NameNotInContext, "y12").WithArguments("y12").WithLocation(98, 29),
+                // (97,17): warning CS0219: The variable 'y12' is assigned but its value is never used
+                //             var y12 = 12;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "y12").WithArguments("y12").WithLocation(97, 17),
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: >Diagnostic [](start = 16, length = 10)

This diagnostic looks like it may be a true breaking change, for code such as
```cs
do
    var x = 1;
while (false);
```
However, I expect it is unlikely to occur in production code.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: To clarify, addition TypeSyntaxBinder property wasn't necessary, addition of initializer binder for LacalWithInitializer was necessary for correct behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 9 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Oh, I see it is preceded by an error for an embedded declaration. Good.

---
In reply to: [75559057](https://github.com/dotnet/roslyn/pull/13271#discussion_r75559057) [](ancestors = 75559057)
